0000-zmix/README.md

```diff
@@ -0,0 +1,395 @@
+- Feature Name: (mal000002, zmix)
+- Start Date: 05-Feb-2019
+- RFC PR: (leave this empty)
+- Ursa Issue: (leave this empty)
+- Version: 1
+
+# Summary
+[summary]: #summary
+
+Zmix is a library for expressing, constructing, and verifying non-interactive
+zero-knowledge proofs (ZKPs). A broad class of zero-knowledge proofs is 
+supported. Schnorr-type proofs are used for many parts and to glue pieces
+together, while individual parts of a proof may use other zero-knowledge
+techniques. 
+
+# Motivation
+[motivation]: #motivation
+
+Zero-knowledge proofs form a cryptographic building block with many 
+applications. Within hyperledger, Indy relies heavily on ZKPs to construct
+anonymous credentials. Fabric uses ZKPs to authenticate transactions when
+using the identity mixer membership service provider. Implementing ZKPs
+correctly however is challenging. Zmix aims to offer a secure
+implementation of ZKPs which can be consumed through a simple API, making
+it easier for other projects to use ZKPs.
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+## Introduction
+
+A zero-knowledge proof enables a prover to convince a verifier of the truth
+of some statement about certain values, without revealing those values. For
+example, a prover may know that values `A` and `B` have an equal discrete
+logarithm. That is, for some `x`, `A = g^x` and `B = h^x`. It could
+convince a verifier that this is true by revealing `x`. If the prover would
+rather not disclose `x`, it could instead construct a zero-knowledge proof
+of this statement, and send the resulting proof to the verifier. The 
+verifier can now be sure that `log_g(A) = log_h(B)`, without knowing the
+discrete log `x`.
+
+In the example above, there are three components: 
+
+- What do we want to prove? That there exists an `x` such that `A = g^x`
+and `B = h^x`.
+
+- How do we know this is true? Because we know `x`.
+
+- How can we prove it? With a zero-knowledge proof that the prover produces
+and the verifier verifies.
+
+In zmix, we have data types corresponding to these three components:
+
+- `ProofSpec`: specifies what we want to prove
+
+- `Witness`: contains the secrets that satisfy the `ProofSpec`.
+
+- `Proof`: the zero-knowledge proof. 
+
+The library provides two features: construction and verification of
+zero-knowledge proofs. 
+
+To construct a proof, zmix requires a `ProofSpec s` and a `Witness w`
+(which is a valid witness for `p`), and will output a `Proof p`. 
+
+To verify a proof, zmix requires a `ProofSpec s` and a `Proof p`, and will
+output a boolean indicating whether the proof was valid (true) or invalid
+(false).
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+Conceptually, the zmix library will offer the functions:
+
+* `generate_proof(s: ProofSpec, w: Witness) -> Result<Proof, Error>`
+* `verify_proof(p: Proof, s: ProofSpec) -> Result<(), Error>`
```

Comment:
This function signature indicates that on failure there is error and success means no error which contradicts the statement given above "To verify a proof, zmix requires a `ProofSpec s` and a `Proof p`, and will output a boolean".